### PR TITLE
Feature/point cloud xyz

### PR DESCRIPTION
**Pull request by lpdon**
Seems to work as expected, tested yesterday with the camera. Please check if the code is ok.

P.S: submitting a PR because I don't have write access. 

pc_xyz_rgb
pc_xyz

### CODE REVIEW

File: realsense2_camera/src/base_realsense_node.cpp [hunk 1]
```diff
@@ -1131,6 +1141,74 @@ void BaseRealSenseNode::publishRgbToDepthPCTopic(const ros::Time& t, const std::
...
+    // Fill the PointCloud2 fields
+    for (int y = 0; y < depth_intrinsics.height; ++y)
+    {
+        for (int x = 0; x < depth_intrinsics.width; ++x)
+        {
+            scaled_depth = static_cast<float>(*image_depth16) * _depth_scale_meters;
+            float depth_pixel[2] = {static_cast<float>(x), static_cast<float>(y)};
+            rs2_deproject_pixel_to_point(depth_point, &depth_intrinsics, depth_pixel, scaled_depth);
+
+            if (depth_point[2] <= 0.f || depth_point[2] > 5.f)
```
Comment (FabianReister) [inline realsense2_camera/src/base_realsense_node.cpp]: Where does 5.0 come from? This should be a parameter. I would also appreciate a MIN_DEPTH parameter with the values from table 4-5 (https://software.intel.com/sites/default/files/Intel_RealSense_Depth_Cam_D400_Series_Datasheet.pdf). 

Furthermore, please see 'figure 4-1 Left Invalid Depth Band' in that PDF. The points for a small x-value are not stable. 


[Copy of https://software.intel.com/sites/default/files/Intel_RealSense_Depth_Cam_D400_Series_Datasheet.pdf](https://github.com/KITcar-Team/realsense/files/2195666/Intel_RealSense_Depth_Cam_D400_Series_Datasheet.pdf)

File: realsense2_camera/src/base_realsense_node.cpp [hunk 2]
```diff
@@ -1131,6 +1141,74 @@ void BaseRealSenseNode::publishRgbToDepthPCTopic(const ros::Time& t, const std::
...
+            {
+                depth_point[0] = 0.f;
```
Comment (FabianReister) [inline realsense2_camera/src/base_realsense_node.cpp]: Invalid points should not be added to the point cloud. 

If you change msg_pointcloud_depth.height to 1 and msg_pointcloud_depth.width to depth_intrinsics.width * depth_intrinsics.height, the unused memory can be easily freed after the loops.

**Comment by lpdon** [comment]
Although this is probably wrong because it just handles the unaligned points...

**Comment by FabianReister** [comment]
@lpdon any better suggestion?

**Comment by lpdon** [comment]
Checking the original repo (https://github.com/intel-ros/realsense) the rs_rgbd.launch publishes the aligned_depth by default. I tested it with enable_pointcloud:=true because I misinterpreted the code as I thought it was the only way to publish the pointcloud, which is indeed published but looks very bad compared to the aligned depth topic: 

35396613-ddcb1d6c-01f5-11e8-8887-4debf178d0cc

I would rather check the topics with pointclouds again with the camera and before merging this.

**Comment by lpdon** [comment]
I spent sometime today with the camera and unterstood how things actually work. The way it was implemented in this PR is not wrong, I would say it is ok as a temporary solution, apart from the "frame" around the pointcloud the data is the same. The thing is: this internal generation of the pointcloud by the realsense ros driver is deprecated and will be removed in the future. The current solution is to call the rgbd nodelet and pass all the raw data. The nodelet then processes and publishes it. 
The right solution that I'll implement is something like: subscribe to the processed depth_registered/points and re-publish them without the color.

**Comment by FabianReister** [comment]
@lpdon we already have a temporary solution. Please take the time you need to implement an efficient solution in terms of runtime and memory usage as our hardware is limited. 

Computing and publishing the XYZRGB point cloud does not seem to be optimal.

**Comment by lpdon** [comment]
I will look somewhere else but it doesn't look like there is much to do directly in the realsense driver as the processing is done by the rgbd nodelet. The optimal solution seems to leave the realsense driver the way it is and change the rgbd.

**Comment by FabianReister** [comment]
Please take a look at the rgbd launch file. You could just change depth_image_proc/point_cloud_xyzrgb to depth_image_proc/point_cloud_xyz and we are done ;)

**Comment by FabianReister** [comment]
If the components are so loosely coupled, we could also store the depth image in the bag file...

**Comment by lpdon** [comment]
"Please take a look at the rgbd launch file. You could just change depth_image_proc/point_cloud_xyzrgb to depth_image_proc/point_cloud_xyz and we are done ;)"

@FabianReister that would be almost too embarrassingly easy :) but you are right the best solution is probably just a rgbd setting. Are there any rosbags containing the raw data published to rgbd? The ones I found on the server have only /camera/color/image_raw and the evil /camera/depth/color/points. I was thinking only of the re-publish and recorded a rosbag with the depth_registered/points :(

**Comment by FabianReister** [comment]
Don't care about the bag files we recorded so far. The camera mount was quite hacky and loose. Therefore, we need to recalibrate. As far as I know there were no bag files with the raw data.

I think we now found a promising solution such that we can extend the rosbag script (depth image + camera info, rgb optional). The old bagfiles should then be deleted. 

You might want to copy the rgbd launch file that will be tailored to our needs. 

We should discuss this in Redmine and create tasks.

**Comment by lpdon** [comment]
@FabianReister spent some more time playing with it... There is indeed something similar to point_cloud_xyz in rgbd. Just called it directly with some new parameters. Compared 2 rosbags rgbd vs depth only and the difference is not bad: 

path:        2018-07-20-14-52-35.bag
version:     2.0
duration:    5.2s
start:       Jul 20 2018 14:52:35.47 (1532091155.47)
end:         Jul 20 2018 14:52:40.69 (1532091160.69)
size:        1.4 GB
messages:    157
compression: none [157/157 chunks]
types:       sensor_msgs/PointCloud2 [1158d486dd51d683ce2f1be655c3c181]
topics:      /camera/depth_registered/points   157 msgs    : sensor_msgs/PointCloud2
---
path:        2018-07-20-14-52-55.bag
version:     2.0
duration:    5.3s
start:       Jul 20 2018 14:52:55.82 (1532091175.82)
end:         Jul 20 2018 14:53:01.12 (1532091181.12)
size:        745.4 MB
messages:    159
compression: none [159/159 chunks]
types:       sensor_msgs/PointCloud2 [1158d486dd51d683ce2f1be655c3c181]
topics:      /camera/depth/points   159 msgs    : sensor_msgs/PointCloud2

Please take a look and say if this is a reasonable solution. 
P.S: as you also said, in the end the rosbags could basically record only the camera raw data and the rgbd nodelet could be called during the replay to generate the pointcloud if needed.

**Review by FabianReister — APPROVED** [review]

File: realsense2_camera/src/base_realsense_node.cpp (hunk 1 above)
Comment (lpdon) [reply realsense2_camera/src/base_realsense_node.cpp]: I reverted these changes (.h and .cpp) in the last commit. The only thing that should be changed is the launch file. Everything related to pointcloud generation shouldn't be done in the realsense driver.

**Pull request merged**